### PR TITLE
(feat) X-Forwarded-For spoofing mitigation

### PR DESCRIPTION
**Pull request by michalc**

### CODE REVIEW

File: activity_stream/views.py [hunk 1]
```diff
@@ -85,28 +85,45 @@ def authenticate(self, request):
...
             logger.warning(
-                'Failed authentication: the X-Forwarded-For header did not '
-                'start with an IP in the whitelist'
+                'Failed authentication: the X-Forwarded-For header was not '
+                'produced by a whitelisted IP'
             )
             raise AuthenticationFailed(INCORRECT_CREDENTIALS_MESSAGE)
 
+    def _authenticate_by_hawk(self, request):
+        print('GETTING TO HAWK')
```
Comment (reupen) [inline activity_stream/views.py]: Did you mean to leave these print() calls in?

Comment (michalc) [reply activity_stream/views.py]: Oops... removed

File: activity_stream/views.py [hunk 2]
```diff
@@ -85,22 +85,36 @@ def authenticate(self, request):
...
         if 'HTTP_X_FORWARDED_FOR' not in request.META:
             logger.warning(
                 'Failed authentication: no X-Forwarded-For header passed'
             )
             raise AuthenticationFailed(INCORRECT_CREDENTIALS_MESSAGE)
 
         x_forwarded_for = request.META['HTTP_X_FORWARDED_FOR']
-        remote_address = x_forwarded_for.split(',', maxsplit=1)[0].strip()
+        ip_addesses = x_forwarded_for.split(',')
+        if len(ip_addesses) < 2:
```
Comment (reupen) [inline activity_stream/views.py]: Can we use a named constant for the 2 here and below?

File: activity_stream/views.py [hunk 3]
```diff
@@ -85,22 +85,36 @@ def authenticate(self, request):
         If either of these suggest we cannot authenticate, AuthenticationFailed
         is raised, as required in the DRF authentication flow
         """
+        self._authenticate_by_ip(request)
```
Comment (reupen) [inline activity_stream/views.py]: The function names are a bit confusing as they sound like two alternative authentication methods – maybe something like _check_ip() and _check_hawk_header() instead?

File: activity_stream/views.py (hunk 1 above)
Comment (stuaxo) [reply activity_stream/views.py]: https://giphy.com/gifs/qVVVfmHDMBZug

File: activity_stream/views.py [hunk 4]
```diff
@@ -85,22 +86,36 @@ def authenticate(self, request):
...
+        ip_addesses = x_forwarded_for.split(',')
+
+        if len(ip_addesses) < PAAS_ADDED_X_FORWARDED_FOR_IPS:
+            logger.warning(
+                'Failed authentication: the X-Forwarded-For header does not '
+                'contain enough IP addresses'
+            )
+            raise AuthenticationFailed(INCORRECT_CREDENTIALS_MESSAGE)
+
+        remote_address = ip_addesses[PAAS_ADDED_X_FORWARDED_FOR_IPS].strip()
```
Comment (reupen) [inline activity_stream/views.py]: Should that be -PAAS_ADDED_X_FORWARDED_FOR_IPS as it was -2 before?

Comment (michalc) [reply activity_stream/views.py]: Fixed!

File: activity_stream/views.py (hunk 3 above)
Comment (michalc) [reply activity_stream/views.py]: Changed!

File: activity_stream/views.py (hunk 2 above)
Comment (michalc) [reply activity_stream/views.py]: Done!

File: activity_stream/views.py (hunk 1 above)
Comment (michalc) [reply activity_stream/views.py]: Removed! (And enjoying the very appropriate gif...)

File: activity_stream/views.py [hunk 5]
```diff
@@ -85,22 +86,36 @@ def authenticate(self, request):
         If either of these suggest we cannot authenticate, AuthenticationFailed
         is raised, as required in the DRF authentication flow
         """
+        self._check_ip(request)
+        return self._check_hawk_header(request)
+
+    def _check_ip(self, request):
         if 'HTTP_X_FORWARDED_FOR' not in request.META:
```
Comment (reupen) [inline activity_stream/views.py]: It's not obvious what this is doing – could you add a docstring to explain it further and what assumptions are being made? Will both PaaS to PaaS and internet to PaaS work? What scenarios will it block?

**Comment by reupen** [comment]
Also would be useful to add something about Hawk to PROJECT_README.md as it would probably be surprising to read that and then encounter Hawk. (Even better would be info on how to run a client and server locally for development :slightly_smiling_face: )

File: activity_stream/views.py (hunk 5 above)
Comment (michalc) [reply activity_stream/views.py]: Have added a bit of info

**Comment by michalc** [comment]
@reupen Have added a bit of documentation with a high-level explanation and what I think are the main gotchas.

**Pull request merged**